### PR TITLE
WIP: get repo on sourcegraph dot com calls repo-lookup so repo gets updated

### PR DESCRIPTION
**Pull request by mollylogue**
Fixes https://github.com/sourcegraph/sourcegraph/issues/25217

### CODE REVIEW

File: cmd/frontend/backend/repos.go [hunk 1]
```diff
@@ -79,12 +79,19 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
 	ctx, done := trace(ctx, "Repos", "GetByName", name, &err)
 	defer done()
 
+	sourceGraphDotComMode := envvar.SourcegraphDotComMode()
+
+	if sourceGraphDotComMode {
+		// TODO: handle the output from this somehow?
+		callRepoUpdaterForRepo(ctx, name)
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:86]: Each call here will end up syncing the repo which leads to a code host API call so we would hit our rate limit pretty quickly.

We do have *some* rate limiting in place but I doubt it would be enough:
https://github.com/sourcegraph/sourcegraph/blob/c0a129d0ba096c1497ae9c9cf1bf37e7ddece0b7/internal/repos/syncer.go#L285

Instead of syncing on every visit, I think we should only try and sync when there might be an issue. We can use errors in the `gitserver_repos` table as a hint.

Comment (mollylogue) [reply cmd/frontend/backend/repos.go:86]: Ah okay great. I figured that would come into play somewhere. I don't think I want to add it inside the syncer, since we do want the syncer to always run when adding a new repo, so maybe that check needs to wrap the call to this function somehow?

File: cmd/frontend/backend/repos.go [hunk 2]
```diff
@@ -103,6 +119,42 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
...
+		// no gitserverRepoStore was initialized, so don't try to update the repo.
+		return
+	}
+	gsRepo, err := s.gitserverReposStore.GetByID(ctx, repoID)
+	if err != nil {
+		// Question: should we attempt to update the repo if we run into an error when fetching it from gitserver_repos?
+		log15.Error(fmt.Sprintf("Error getting repo from gitserver_repos: %s", err))
+		return
+	}
+	log15.Info("Successfully fetched from gs repo store")
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:137]: I think should be at debug level or it could be pretty noisy. Best to only log problems, not successes.

Another option would be to instrument this instead.

File: cmd/frontend/backend/repos.go [hunk 3]
```diff
@@ -103,6 +119,42 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
...
+// gitserverRepoStore was never initialized by just silently passing through.
+func (s *repos) checkAndUpdateRepo(ctx context.Context, repoID api.RepoID, name api.RepoName) {
+	if s.gitserverReposStore == nil {
+		// no gitserverRepoStore was initialized, so don't try to update the repo.
+		return
+	}
+	gsRepo, err := s.gitserverReposStore.GetByID(ctx, repoID)
+	if err != nil {
+		// Question: should we attempt to update the repo if we run into an error when fetching it from gitserver_repos?
+		log15.Error(fmt.Sprintf("Error getting repo from gitserver_repos: %s", err))
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:134]:
```suggestion
		log15.Error("Error getting repo from gitserver_repos", "error", err))
```

We usually append a list of key value pairs when using log15

File: cmd/frontend/backend/repos.go [hunk 4]
```diff
@@ -103,6 +119,42 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
...
+		return
+	}
+	log15.Info("Successfully fetched from gs repo store")
+	if gsRepo != nil && gsRepo.LastError != "" {
+		// LastError is not empty, which means the last time we tried to update the repo there was an error.
+		// In this case, we should call out to the repo-updater to update the repo.
+		_, err := callRepoUpdaterForRepo(ctx, name)
+		if err != nil {
+			log15.Error(fmt.Sprintf("Error updating repo: %s", err))
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:143]:
```suggestion
			log15.Error(fmt.Sprintf("Error updating repo", "error", err))
```

File: cmd/frontend/backend/repos.go [hunk 5]
```diff
@@ -103,6 +119,42 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
...
+		}
+		log15.Info(fmt.Sprintf("last error for repo %s was non empty: (%s) so the repo was updated",
+			name, gsRepo.LastError))
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:146]:
```suggestion
		log15.Info(fmt.Sprintf("last error for repo was non empty, so the repo was updated",
			"repo", name, "error", gsRepo.LastError))
```

File: internal/repos/syncer.go [hunk 6]
```diff
@@ -294,6 +294,9 @@ func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName) (repo *types.R
 				if err != nil {
 					log15.Error("Error syncing repo in the background", "name", name, "error", err)
 				}
+				// TODO: When calling this on a repo that's been recently made private from public, this error is
```
Comment (ryanslade) [inline internal/repos/syncer.go:297]: Agreed, but happy to leave for another PR

File: internal/extsvc/github/v3.go [hunk 7]
```diff
@@ -441,7 +441,7 @@ func (c *V3Client) GetRepository(ctx context.Context, owner, name string) (*Repo
 			keys = append(keys, nodeIDCacheKey(repo.ID)) // also cache under GraphQL node ID
 		}
 		return repo, keys, err
-	}, false)
+	}, true)
```
Comment (ryanslade) [inline internal/extsvc/github/v3.go:444]: I assume we don't want to leave this in?

File: cmd/frontend/backend/repos.go [hunk 8]
```diff
@@ -103,6 +119,42 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
...
+	}
+}
+
+func callRepoUpdaterForRepo(ctx context.Context, name api.RepoName) (api.RepoName, error) {
+	lookupResult, err := repoupdater.DefaultClient.RepoLookup(ctx, protocol.RepoLookupArgs{Repo: name})
```
Comment (ryanslade) [inline cmd/frontend/backend/repos.go:151]: I think there's still a danger that we hit the github api too often here so I think we have two options: 

1. Make this a no-op initially and just instrument instead so we can observe how often it's going to get called in practice
2. Add a simple in memory rate limiter here set to something low like 100 repos per hour or something so that over time we slowly fix these repos but don't run the risk of consuming all our tokens. Also, bear in mind that the rate limiter would be in memory so it would be per frontend instance (currently 4)

**Comment by ryanslade** [comment]
Nice! This is looking great so far 👍 I didn't realise we'd have to do all that database mocking work too which unfortunately makes it a bigger change

**Comment by tsenart** [comment]
It would be great to have a detailed PR description :) Not sure I understand the why of this change as it stands.

**Comment by ryanslade** [comment]
> It would be great to have a detailed PR description :) Not sure I understand the why of this change as it stands.

I added a link to the related issue

File: internal/extsvc/github/v3.go (hunk 7 above)
Comment (mollylogue) [reply internal/extsvc/github/v3.go:444]: Nope! It was just for testing purposes. It took me and Indra nearly an hour to figure out why the behavior was so weird and non-deterministic and then we realized we were caching the github calls 🤦‍♀️ So I switched it off for testing and definitely didn't mean to commit it.

**Comment by mollylogue** [comment]
> It would be great to have a detailed PR description :) Not sure I understand the why of this change as it stands.

This PR is a work in progress and mostly only posted so people could give me thoughts since most of my team has very little overlap with me timezone-wise. I'll likely add a description and remove the "WIP" from the title today once I think it's ready for a formal review.

File: cmd/frontend/backend/repos.go (hunk 8 above)
Comment (mollylogue) [reply cmd/frontend/backend/repos.go:151]: Okay! I'm down to do the in-memory rate-limiting. Maybe adding some instrumentation around it as well would be a smart idea so we get some visibility into how often this is happening and how often we hit the rate limiter.

**Comment by tsenart** [comment]
> This PR is a work in progress and mostly only posted so people could give me thoughts since most of my team has very little overlap with me timezone-wise. I'll likely add a description and remove the "WIP" from the title today once I think it's ready for a formal review.

No worries! I just got notified because I'm subscribed to changes in `internal/repos/syncer.go`, so I was curious. I missed the WIP prefix, so sorry about that :)

**Comment by mollylogue** [comment]
Closing this as we're opting for an approach where we sync these problematic repos in the background. https://github.com/sourcegraph/sourcegraph/pull/28522
